File: evidence/20260917-design-tests/new-tests/VehicleDesignBoundaryTest.java
package com.example.carlauncher.model;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import java.util.Arrays;
import java.util.Collection;
import static org.junit.Assert.assertEquals;

/** Design section 6.2: rounding precedes range validation; speed errors precede gear errors. */
@RunWith(Parameterized.class)
public class VehicleDesignBoundaryTest {
    @Parameterized.Parameters(name = "{index}: speed={0}, gear={1}, validity={2}")
    public static Collection<Object[]> cases() {
        return Arrays.asList(new Object[][] {
                {-0.5f, Gear.D, DataValidity.VALID},
                {Math.nextDown(-0.5f), Gear.D, DataValidity.INVALID_SPEED},
                {0f, Gear.P, DataValidity.VALID},
                {200f, Gear.D, DataValidity.VALID},
                {Math.nextDown(200.5f), Gear.D, DataValidity.VALID},
                {200.5f, Gear.D, DataValidity.INVALID_SPEED},
                {Float.MAX_VALUE, Gear.D, DataValidity.INVALID_SPEED},
                {-Float.MAX_VALUE, Gear.D, DataValidity.INVALID_SPEED},
                {Float.NaN, null, DataValidity.INVALID_SPEED},
                {Float.POSITIVE_INFINITY, null, DataValidity.INVALID_SPEED},
                {0f, null, DataValidity.INCOMPLETE}
        });
    }

    private final float speed;
    private final Gear gear;
    private final DataValidity expected;

    public VehicleDesignBoundaryTest(float speed, Gear gear, DataValidity expected) {
        this.speed = speed;
        this.gear = gear;
        this.expected = expected;
    }

    @Test public void validate_respectsRoundedBoundaryAndErrorPrecedence() {
        assertEquals(expected, VehicleStateValidator.validate(speed, gear));
    }
}

File: app/src/test/java/com/example/carlauncher/model/VehicleDesignBoundaryTest.kt
package com.example.carlauncher.model

import java.util.Arrays
import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized

/* Design section 6.2: rounding precedes range validation; speed errors precede gear errors. */
@RunWith(Parameterized::class)
open class VehicleDesignBoundaryTest(
    private val speed: Float,
    private val gear: Gear?,
    private val expected: DataValidity?,
) {

    @Test
    open fun validate_respectsRoundedBoundaryAndErrorPrecedence() {
        assertEquals(expected, VehicleStateValidator.validate(speed, gear))
    }

    companion object {
        @Parameterized.Parameters(name = "{index}: speed={0}, gear={1}, validity={2}")
        @JvmStatic
        fun cases(): Collection<Array<Any?>?>? =
            Arrays.asList<Array<Any?>?>(
                *arrayOf<Array<Any?>?>(
                    arrayOf<Any?>(-0.5f, Gear.D, DataValidity.VALID),
                    arrayOf<Any?>(Math.nextDown(-0.5f), Gear.D, DataValidity.INVALID_SPEED),
                    arrayOf<Any?>(0f, Gear.P, DataValidity.VALID),
                    arrayOf<Any?>(200f, Gear.D, DataValidity.VALID),
                    arrayOf<Any?>(Math.nextDown(200.5f), Gear.D, DataValidity.VALID),
                    arrayOf<Any?>(200.5f, Gear.D, DataValidity.INVALID_SPEED),
                    arrayOf<Any?>(java.lang.Float.MAX_VALUE, Gear.D, DataValidity.INVALID_SPEED),
                    arrayOf<Any?>(-java.lang.Float.MAX_VALUE, Gear.D, DataValidity.INVALID_SPEED),
                    arrayOf<Any?>(java.lang.Float.NaN, null, DataValidity.INVALID_SPEED),
                    arrayOf<Any?>(
                        java.lang.Float.POSITIVE_INFINITY,
                        null,
                        DataValidity.INVALID_SPEED,
                    ),
                    arrayOf<Any?>(0f, null, DataValidity.INCOMPLETE),
                )
            )
    }
}
